### ISSUE TITLE
AbstractAssert.isInstanceOf(Class<?>) should return a narrowed assert object

### ISSUE THREAD

**Issue by JnRouvignac** [joel-costigliola/assertj-core#1131]
#### Summary

`isInstanceOf(Class<?> type)` helps a lot writing fluent tests, but it would be even more useful if it could cast the returned instance to be the same as the provided type. The same applies to `isExactlyInstanceOf(Class<?> type)`.

#### Example

Here is some starting code:
```java
assertThatThrownBy(() -> searchEntry(TEST_ROOT_DN, filterForBadValue))
    .isInstanceOf(LdapException.class)
    .extracting(e -> ((LdapException) e).getResultCode())
    .isEqualTo(CLIENT_SIDE_NO_RESULTS_RETURNED));
```

It would be really nice if it could be rewritten as:

```java
assertThatThrownBy(() -> searchEntry(TEST_ROOT_DN, filterForBadValue))
    .isInstanceOf(LdapException.class)
    .extracting(e -> e.getResultCode())
    .isEqualTo(CLIENT_SIDE_NO_RESULTS_RETURNED);
```

Note that I am not super satisfied (ha ha! pun intended) with the alternative which feels heavier and harder to read IMO (and it is harder to fit on a single line):

```java
assertThatThrownBy(() -> searchSingleEntry(TEST_ROOT_DN, filterForBadValue))
        .isInstanceOfSatisfying(LdapException.class,
                                e -> assertThat(e.getResultCode()).isEqualTo(CLIENT_SIDE_NO_RESULTS_RETURNED));
```

#### Java 8 specific ?

* NO  : create Pull Request from the `2.x` branch

However, I think this is breaking compatibility, so it may need to be pushed to a new major release?

**Comment by joel-costigliola** [comment]
I'm afraid this is not technically possible but I'm happy to be proven wrong (see https://github.com/joel-costigliola/assertj-core/issues/683).

The next version will provide `catchThrowableOfType` that can catch an exception and cast it to a specific type, this should cater for for use case.

In the meantime you can give a try to [`isInstanceOfSatisfying`](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractAssert.html#isInstanceOfSatisfying-java.lang.Class-java.util.function.Consumer-).

**Comment by JnRouvignac** [comment]
Ok, thanks for letting me know, I did not find it by myself.
As mentioned I have tried `isInstanceOfSatisfying()`, which is sometimes helpful and sometimes not what I expect.

But this is a complaint from a rich man :)
AssertJ is extremely powerful and gets so far as to help having very useful assertion failure messages.
Thanks a lot for maintaining it, as soon as I used it, I knew that I had been missing such a library for years.

I can't wait to put my hand on `catchThrowableOfType()`.
In the meantime, I'll close this issue.
Thanks again!
